cvat/apps/engine/serializers.py

```diff
@@ -1980,63 +1982,55 @@ class Meta:
+    instance: Union[models.Project, models.Task],
+    validated_data: Dict[str, Any],
+) -> None:
+    for storage_type in ('source_storage', 'target_storage'):
+        new_conf = validated_data.pop(storage_type, None)
 
         if not new_conf:
             continue
 
-        cloud_storage_id = new_conf.get('cloud_storage_id')
+        cloud_storage, cloud_storage_id = None, new_conf.get('cloud_storage_id')
         if cloud_storage_id:
-            _validate_existence_of_cloud_storage(cloud_storage_id)
+            try:
+                cloud_storage = models.CloudStorage.objects.get(id=cloud_storage_id)
+            except models.CloudStorage.DoesNotExist:
+                raise serializers.ValidationError(f'The specified cloud storage {cloud_storage_id} does not exist.')
 
         # storage_instance maybe None
-        storage_instance = getattr(instance, storage)
+        storage_instance = getattr(instance, storage_type)
         if not storage_instance:
             storage_instance = models.Storage(**new_conf)
             storage_instance.save()
-            setattr(instance, storage, storage_instance)
+            setattr(instance, storage_type, storage_instance)
             continue
 
-        new_location = new_conf.get('location')
-        storage_instance.location = new_location or storage_instance.location
-        storage_instance.cloud_storage_id = new_conf.get('cloud_storage_id', \
-            storage_instance.cloud_storage_id if not new_location else None)
-
-        cloud_storage_id = storage_instance.cloud_storage_id
-        if cloud_storage_id:
-            try:
-                _ = models.CloudStorage.objects.get(id=cloud_storage_id)
-            except models.CloudStorage.DoesNotExist:
-                raise serializers.ValidationError(f'The specified cloud storage {cloud_storage_id} does not exist.')
-
+        new_location = new_conf.get('location', storage_instance.location)
+        location_has_changed = new_location != storage_instance.location
+        storage_instance.location = new_location
+        storage_instance.cloud_storage = cloud_storage if cloud_storage or location_has_changed else storage_instance.cloud_storage
         storage_instance.save()
 
-def _configure_related_storages(validated_data):
-
+def _configure_related_storages(validated_data: Dict[str, Any]) -> Dict[str, Optional[models.Storage]]:
     storages = {
         'source_storage': None,
         'target_storage': None,
     }
 
     for i in storages:
-        storage_conf = validated_data.get(i)
-        if storage_conf:
-            cloud_storage_id = storage_conf.get('cloud_storage_id')
-            if cloud_storage_id:
-                _validate_existence_of_cloud_storage(cloud_storage_id)
+        if storage_conf := validated_data.get(i):
+            if cloud_storage_id := storage_conf.get('cloud_storage_id'):
+                try:
+                    _ = models.CloudStorage.objects.get(id=cloud_storage_id)
```

Comment:
How about using `.filter(id=cloud_storage_id).exists()` to avoid fetching the actual object?